### PR TITLE
chore: new error message

### LINKED ISSUE #120: Confusing error message with partial and inner function
### What I Did

```python
from in_n_out import Store

store = Store.create('my-store')

@store.inject
def myfun(a: str):
    def inner_fun(thing: 'thing'):
        return thing
    return inner_fun()

store.inject(partial(myfun, a='a'))()
```

gives the following error message:
```
TypeError: After injecting dependencies for NO arguments, After injecting dependencies for NO arguments, myfun.<locals>.inner_fun() missing 1 required positional argument: 'thing'
```

<details>

<summary>Full error message</summary>

```
Traceback (most recent call last):
  File "/home/lucy/Documents/dev/in-n-out/src/in_n_out/_store.py", line 805, in _exec
    result = func(**bound.arguments)
  File "<ipython-input-9-e899288c75f6>", line 5, in myfun
    return inner_fun()
TypeError: myfun.<locals>.inner_fun() missing 1 required positional argument: 'thing'
After injecting dependencies for NO arguments, myfun.<locals>.inner_fun() missing 1 required positional argument: 'thing'
Traceback (most recent call last):
  File "/home/lucy/Documents/dev/in-n-out/src/in_n_out/_store.py", line 805, in _exec
    result = func(**bound.arguments)
  File "<ipython-input-9-e899288c75f6>", line 5, in myfun
    return inner_fun()
TypeError: myfun.<locals>.inner_fun() missing 1 required positional argument: 'thing'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/lucy/Documents/dev/in-n-out/src/in_n_out/_store.py", line 805, in _exec
    result = func(**bound.arguments)
  File "/home/lucy/Documents/dev/in-n-out/src/in_n_out/_store.py", line 828, in _exec
    raise TypeError(
TypeError: After injecting dependencies for NO arguments, myfun.<locals>.inner_fun() missing 1 required positional argument: 'thing'
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~/Documents/dev/in-n-out/src/in_n_out/_store.py:805, in Store.inject.<locals>._inner.<locals>._exec(*args, **kwargs)
    804 try:
--> 805     result = func(**bound.arguments)
    806 except TypeError as e:

Cell In[9], line 5, in myfun(a)
      4     return thing
----> 5 return inner_fun()

TypeError: myfun.<locals>.inner_fun() missing 1 required positional argument: 'thing'

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
File ~/Documents/dev/in-n-out/src/in_n_out/_store.py:805, in Store.inject.<locals>._inner.<locals>._exec(*args, **kwargs)
    804 try:
--> 805     result = func(**bound.arguments)
    806 except TypeError as e:

File ~/Documents/dev/in-n-out/src/in_n_out/_store.py:828, in Store.inject.<locals>._inner.<locals>._exec(*args, **kwargs)
    822             logger.error(
    823                 f"Do not have argument for {param.name}: using "
    824                 "providers "
    825                 f"{list(self.iter_providers(param.annotation))}"
    826             )
--> 828     raise TypeError(
    829         f"After injecting dependencies for {_argnames}, {e}"
    830     ) from e
    832 return result

TypeError: After injecting dependencies for NO arguments, myfun.<locals>.inner_fun() missing 1 required positional argument: 'thing'

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
Cell In[13], line 1
----> 1 store.inject(partial(myfun, a='a'))()

File ~/Documents/dev/in-n-out/src/in_n_out/_store.py:828, in Store.inject.<locals>._inner.<locals>._exec(*args, **kwargs)
    818         if (
    819             param.name not in bound.arguments
    820             and param.default is param.empty
    821         ):
    822             logger.error(
    823                 f"Do not have argument for {param.name}: using "
    824                 "providers "
    825                 f"{list(self.iter_providers(param.annotation))}"
    826             )
--> 828     raise TypeError(
    829         f"After injecting dependencies for {_argnames}, {e}"
    830     ) from e
    832 return result
```

</details>

This is confusing as in-n-out has not attempted to injection on the inner function, but the error message makes it seem like it has. As a user, I think the error message should just be the original, no mention of injecting dependencies, but @dragadoncila can weigh in as well.

Note the `partial`, which I think is the part that causes in-n-out to try the injecting.

This example may seem contrived but was experienced when trying to add a widget like below in napari:

```python
import napari

from qtpy.QtWidgets import QVBoxLayout, QWidget
from magicgui.widgets import create_widget

def create_combined_widget(viewer: napari.viewer.Viewer):

    main_widget = QWidget()
    layout = QVBoxLayout(main_widget)

    data_widget = create_widget(label='Threshold', annotation=float, widget_type='FloatSlider')
    data_widget.changed.connect(lambda x: print(viewer.layers))

    layout.addWidget(data_widget.native)
    return main_widget
```

Note widget callback is [`_toggle_or_get_widget`](https://github.com/napari/napari/blob/33da488d45441fabf485b402dbe6e98c62957769/napari/_qt/_qplugins/_qnpe2.py#L323), in a `partial`.

### PR DESCRIPTION
**Pull request by tlambert03**
closes #120 

@lucyleeow, running your example now would show this:

```python
from in_n_out import Store
from functools import partial

store = Store.create('my-store')

@store.inject
def myfun(a: str):
    def inner_fun(thing: 'thing'):
        return thing
    return inner_fun()

store.inject(partial(myfun, a='a'))()
```

```pytb
Traceback (most recent call last):
  File "/Users/talley/dev/self/in-n-out/src/in_n_out/_store.py", line 801, in _exec
    result = func(**bound.arguments)  # type: ignore[call-arg]
  File "/Users/talley/dev/self/in-n-out/x.py", line 10, in myfun
    return inner_fun()
TypeError: inner_fun() missing 1 required positional argument: 'thing'
Traceback (most recent call last):
  File "/Users/talley/dev/self/in-n-out/src/in_n_out/_store.py", line 801, in _exec
    result = func(**bound.arguments)  # type: ignore[call-arg]
  File "/Users/talley/dev/self/in-n-out/x.py", line 10, in myfun
    return inner_fun()
TypeError: inner_fun() missing 1 required positional argument: 'thing'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/talley/dev/self/in-n-out/x.py", line 12, in <module>
    store.inject(partial(myfun, a='a'))()
  File "/Users/talley/dev/self/in-n-out/src/in_n_out/_store.py", line 801, in _exec
    result = func(**bound.arguments)  # type: ignore[call-arg]
  File "/Users/talley/dev/self/in-n-out/src/in_n_out/_store.py", line 822, in _exec
    raise TypeError(
TypeError: Error calling in-n-out injected function '__main__.myfun' with kwargs {'a': 'a'}.

See TypeError above for more details.
```

more than happy to change it, or if you guys have a more specific idea about what you'd like the error to look like, happy to take PRs and close this one.

note: part of the weird duplication in this error message is caused by calling store.inject on a partial that wraps an already-injected function :) ... so reconsider that pattern too if you can

### CODE REVIEW

**Comment by lucyleeow** [comment]
Thanks @tlambert03, I'm happy with this.

> note: part of the weird duplication in this error message is caused by calling store.inject on a partial that wraps an already-injected function :) ... so reconsider that pattern too if you can

Any suggestion as to how we should go about this instead? We wrap sample and widget plugin contributions in another function, passing args via partial. What would be a better way to do this? :pray:

**Comment by tlambert03** [comment]
Can you point me to the exact code where you do this?

**Comment by lucyleeow** [comment]
For widgets [here](https://github.com/napari/napari/blob/7e27fa80a9383f5b31a4e64a8adb97f18da5cfe4/napari/_qt/_qplugins/_qnpe2.py#L395-L400) is the partial we create for each widget contribution.

For samples menu, [here](https://github.com/napari/napari/blob/7e27fa80a9383f5b31a4e64a8adb97f18da5cfe4/napari/_qt/_qplugins/_qnpe2.py#L257-L261) is the partial we create for each sample contribution.

**Comment by tlambert03** [comment]
ok, i see... but you don't do there what you did in the original example: calling `store.inject()` on a partial that itself wraps another already-injected function right?

```python
@store.inject
def myfun(a: str):
    def inner_fun(thing: 'thing'):
        return thing
    return inner_fun()

store.inject(partial(myfun, a='a'))()
```

(which would be the same as this without the decorator):

```python
def myfun(a: str):
    def inner_fun(thing: 'thing'):
        return thing
    return inner_fun()

# note the double injection
store.inject(partial(store.inject(myfun), a='a'))()
```

maybe that was just an unrelated thing in the example?

**Comment by DragaDoncila** [comment]
@tlambert03 thanks for this change, I do think this error is clearer now.

I'm still a little confused as to whether or not `in-n-out` is actually **attempting** dependency injection in this case (which I think is closest to what is happening in napari):

```python
def myfun(a: str):
    def inner_fun(thing: 'thing'):
        return thing
    return inner_fun()

store.inject(partial(myfun, a='a'))()
```

In this case, all arguments to `myfun` are satisfied before we try to run the function, so there should be no attempt to inject anything right? If nothing is attempted to be injected, then I think maybe this error message, `Error calling in-n-out injected function '__main__.myfun' with kwargs {'a': 'a'}` is still a bit confusing, because it's semi-implying that `in-n-out` satisfied `a`? But maybe it's doesn't imply that and I'm just confused.

TLDR: I think if we try to execute a function from the store, and all arguments to that function are already satisfied (regardless of what annotations are on the function), `in-n-out` should "bail" from the injecting code and just run the function "normally" - as though it was a function with no arguments. Is that what currently happens? Let me know if that question is not clear...

**Comment by lucyleeow** [comment]
Yeah the decorator was errant, I was playing around with the code to try and get a min reproducible example, and didn't delete.

Currently with widgets we say they can be "functions that return a widget instance". We support magic factory functions, BUT we don't actually support a function that returns a widgets instance, e.g., (very pseudo code):

```
def my_widget():
    wid = QWidget()
    return wid()
```

When checking if this worked, we noticed the error above.

Do you think the `partial` structure we have for widgets and sample data is not ideal?

**Comment by tlambert03** [comment]
> I'm still a little confused as to whether or not in-n-out is actually attempting dependency injection in this case (which I think is closest to what is happening in napari):

ahh... ok, i think this confusion helps explain my confusion :)
So, if I'm not mistaken, you're ultimately just calling `Action(callback=partial(...))` right?  And it's app-model itself that's doing the injection.  and so it all probably just happens once right?

> in-n-out should "bail" from the injecting code and just run the function "normally"

yeah, that's correct, there's nothing logically problematic with an "unnecessary" injection.  it just won't do anything... but it *would* have created what appeared to be an unnecessarily longer recursive error message *if* it was a partial around an already-injected partial...  which is why I was just saying that it could be "part of the weird error message".... buuut i think we should just drop all this cause it sounds like @lucyleeow clarified that we can ignore that decorator :)

Anyway,  I think we're all good here.  I don't think there's anything wrong with using a partial.  I might need to chase through the code again a bit to have a clearer answer, but if it's doing what you expect i don't think we have anything to worry about.

**Comment by lucyleeow** [comment]
If I am following correctly, do you think after this part:

https://github.com/pyapp-kit/in-n-out/blob/ad21a896e9f4563d482ea1f701941808628e3a1e/src/in_n_out/_store.py#L772-L773

If all parameters are satisfied, we could bail and just run the func normally (like what we do when there are no annotations)?

**Comment by tlambert03** [comment]
> If all parameters are satisfied, we could bail and just run the func normally

yes, but that's pretty much what the next bit of code checks for (i.e. "whether all parameters are satisfied"):

https://github.com/pyapp-kit/in-n-out/blob/ad21a896e9f4563d482ea1f701941808628e3a1e/src/in_n_out/_store.py#L772-L781

if they're all satasfied, that if clause will never be `True` and it's a no-op, and we just move on to calling `func(**bound.arguments)`

**Comment by lucyleeow** [comment]
I think what @DragaDoncila was wondering is, if all parameters are satisfied, could/should we just do `return func` like we do here:

https://github.com/pyapp-kit/in-n-out/blob/ad21a896e9f4563d482ea1f701941808628e3a1e/src/in_n_out/_store.py#L736-L739

But I am also happy with:

> there's nothing logically problematic with an "unnecessary" injection

**Comment by tlambert03** [comment]
ah! yes it's possible we could bail sooner.  I probably wasn't anticipating that someone might try to inject a function that didn't need any injection.  that would be a good thing to double check and avoid 👍

open issue please?
